src/me.rs

```diff
@@ -731,91 +731,141 @@ pub fn estimate_motion_ss4<T: Pixel>(
+      let mut best_mv = MotionVector::default();
 
-    // Divide by 4 to account for subsampling, 0.125 is a fudge factor
-    let lambda = (fi.me_lambda * 256.0 / 4.0 * 0.125) as u32;
-
-    if fi.config.speed_settings.diamond_me {
-      for omv in pmvs.iter() {
-        if let Some(pmv) = omv {
-          let mut predictors = get_subset_predictors::<T>(
-            &bo_adj_h,
-            MotionVector{row: pmv.row, col: pmv.col},
-            fi.w_in_b, fi.h_in_b,
-            &frame_mvs, frame_ref, 0
-          );
+      // Divide by 4 to account for subsampling, 0.125 is a fudge factor
+      let lambda = (fi.me_lambda * 256.0 / 4.0 * 0.125) as u32;
 
-          for predictor in &mut predictors {
-            predictor.row >>= 1;
-            predictor.col >>= 1;
-          }
-
-          diamond_me_search(
-            fi, &po,
-            &fs.input_hres, &rec.input_hres,
-            &predictors, fi.sequence.bit_depth,
-            global_mv, lambda,
-            mvx_min >> 1, mvx_max >> 1, mvy_min >> 1, mvy_max >> 1,
-            blk_w >> 1, blk_h >> 1,
-            &mut best_mv, &mut lowest_cost,
-            &mut None, 0
-          );
-        }
-      }
+      Self::me(
+        fi, fs, pmvs, bo_adj_h,
+        frame_mvs, frame_ref_opt, po, rec, &global_mv, lambda,
+        mvx_min, mvx_max, mvy_min, mvy_max, blk_w, blk_h,
+        &mut best_mv, &mut lowest_cost
+      );
+
+      Some(MotionVector { row: best_mv.row * 2, col: best_mv.col * 2 })
     } else {
-      for omv in pmvs.iter() {
-        if let Some(pmv) = omv {
-          let x_lo = po.x + (((pmv.col as isize / 8 - range).max(mvx_min / 8).min(mvx_max / 8)) >> 1);
-          let x_hi = po.x + (((pmv.col as isize / 8 + range).max(mvx_min / 8).min(mvx_max / 8)) >> 1);
-          let y_lo = po.y + (((pmv.row as isize / 8 - range).max(mvy_min / 8).min(mvy_max / 8)) >> 1);
-          let y_hi = po.y + (((pmv.row as isize / 8 + range).max(mvy_min / 8).min(mvy_max / 8)) >> 1);
-          full_search(
-            x_lo,
-            x_hi,
-            y_lo,
-            y_hi,
-            blk_h >> 1,
-            blk_w >> 1,
-            &fs.input_hres,
-            &rec.input_hres,
-            &mut best_mv,
-            &mut lowest_cost,
-            &po,
-            1,
-            fi.sequence.bit_depth,
-            lambda,
-            [MotionVector::default(); 2],
-            fi.allow_high_precision_mv
-          );
+      None
+    }
+  }
+
+  fn me<T: Pixel>(
+    fi: &FrameInvariants<T>, fs: &FrameState<T>,
+    pmvs: &[Option<MotionVector>; 3], bo_adj_h: BlockOffset,
+    frame_mvs: &[MotionVector], frame_ref_opt: &Option<Arc<ReferenceFrame<T>>>,
+    po: PlaneOffset, rec: &Arc<ReferenceFrame<T>>,
+    global_mv: &[MotionVector; 2], lambda: u32,
+    mvx_min: isize, mvx_max: isize, mvy_min: isize, mvy_max: isize,
+    blk_w: usize, blk_h: usize,
+    best_mv: &mut MotionVector, lowest_cost: &mut u64
+  );
+}
+
+pub struct DiamondSearchSS2 {}
```

Comment:
We could keep all in a single trait and struct placeholder maybe?

Reply:
Sure.
This is in the last version of the PR.